File: core/src/main/java/net/particify/arsnova/core/validation/LanguageIso639Validator.java
package net.particify.arsnova.core.validation;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import java.util.Locale;
import java.util.Set;
import java.util.stream.Collectors;

public class LanguageIso639Validator implements ConstraintValidator<LanguageIso639, String> {
  private static final Set<String> languages = Set.of(Locale.getISOLanguages()).stream()
      .filter(l -> Locale.of(l).getLanguage().equals(l))
      .collect(Collectors.toSet());

  @Override
  public void initialize(final LanguageIso639 constraintAnnotation) {
    ConstraintValidator.super.initialize(constraintAnnotation);
  }

  @Override
  public boolean isValid(final String value, final ConstraintValidatorContext context) {
    return value == null || languages.contains(value);
  }
}

File: core4/src/main/kotlin/net/particify/arsnova/core4/common/LanguageIso639Validator.kt
/* Copyright 2025 Particify GmbH
 * SPDX-License-Identifier: MIT
 */
package net.particify.arsnova.core4.common

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import java.util.Locale

open class LanguageIso639Validator : ConstraintValidator<LanguageIso639, String?> {
  override fun isValid(value: String?, context: ConstraintValidatorContext): Boolean {
    return value == null || languages.contains(value)
  }

  companion object {
    private val languages =
        Locale.getISOLanguages().filter { it == Locale.of(it).language }.toMutableSet()
  }
}
